packages/@aws-cdk/aws-applicationsignals-alpha/lib/slo/slo-types.ts

```diff
@@ -0,0 +1,138 @@
+import { ISlo, PeriodBasedSloProps, RequestBasedSloProps } from './slo';
+import { ComparisonOperator, MetricType } from './constants';
+import { Goal } from './interval';
+import { PeriodBasedAppSignalsMetricProps, PeriodBasedCloudWatchMetricProps, RequestBasedAppSignalsMetricProps, RequestBasedCloudWatchMetricProps } from './metric';
+import { Resource } from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+import * as applicationsignals from 'aws-cdk-lib/aws-applicationsginals';
+
+/**
+ * Base class for all SLO implementations
+ */
+export abstract class ServiceLevelObjective extends Resource implements ISlo {
+    public readonly arn: string;
+    public readonly name: string;
+    public readonly goal: Goal;
+
+    protected constructor(
+        scope: Construct,
+        id: string,
+        protected readonly props: PeriodBasedSloProps | RequestBasedSloProps
+    ) {
+        super(scope, id);
+        this.name = props.name;
+        this.goal = props.goal;
+    }
+
+    public static periodBased(
+        scope: Construct,
+        id: string,
+        props: PeriodBasedSloProps
+    ): ServiceLevelObjective {
+        return new PeriodBasedSlo(scope, id, props);
+    }
+
+    public static requestBased(
+        scope: Construct,
+        id: string,
+        props: RequestBasedSloProps
+    ): ServiceLevelObjective {
+        return new RequestBasedSlo(scope, id, props);
+    }
+
+    protected createBurnRateConfigurations(): applicationsignals.CfnServiceLevelObjective.BurnRateConfigurationProperty[] | undefined {
+        return this.props.burnRateWindows?.map(minutes => ({
+            lookBackWindowMinutes: minutes,
+        }));
+    }
+
+    protected getMetricComparisonOperator(metric: PeriodBasedAppSignalsMetricProps | PeriodBasedCloudWatchMetricProps | RequestBasedAppSignalsMetricProps | RequestBasedCloudWatchMetricProps): ComparisonOperator {
```

Comment:
if it will be internal function, so it is Ok to use union, but if the plan is to export it to function, so replace the union type with a parent type for all of these concrete types